src/main/java/com/google/ehub/data/FavoriteItemDatastore.java

```diff
@@ -0,0 +1,119 @@
+package com.google.ehub.data;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Singleton class that manages the addition and querying of FavoriteItems stored in Datastore.
+ */
+public final class FavoriteItemDatastore {
+  private static final String FAVORITE_ITEM_KIND = "favoriteItem";
+  private static final String USER_EMAIL_PROPERTY_KEY = "userEmail";
+  private static final String ITEM_ID_PROPERTY_KEY = "itemId";
+
+  private static FavoriteItemDatastore instance;
+
+  private final DatastoreService datastoreService = DatastoreServiceFactory.getDatastoreService();
+
+  private FavoriteItemDatastore() {}
+
+  /**
+   * Gives access to the single instance of the class, and creates this instance
+   * if it was not initialized previously.
+   *
+   * @return single instance of the class
+   */
+  public static FavoriteItemDatastore getInstance() {
+    if (instance == null) {
+      instance = new FavoriteItemDatastore();
+    }
+
+    return instance;
+  }
+
+  /**
+   * Adds a favorite item Entity to Datastore if it doesn't exist yet.
+   *
+   * @param userEmail The email of the user that liked the item
+   * @param itemId the id of the item that was liked
+   */
+  public void addFavoriteItem(String userEmail, Long itemId) {
+    if (doesFavoriteItemExist(userEmail, itemId)) {
+      return;
+    }
+
+    Entity favoriteItemEntity = new Entity(FAVORITE_ITEM_KIND);
+    favoriteItemEntity.setProperty(USER_EMAIL_PROPERTY_KEY, userEmail);
+    favoriteItemEntity.setProperty(ITEM_ID_PROPERTY_KEY, itemId);
+
+    datastoreService.put(favoriteItemEntity);
```

Comment:
Typically when you are accessing a database multiple times (e.g. read, then write) as a single action, you should use a transaction. This prevents the item from being modified by another user/process in between your calls to getItem() and put(). 

Here is a good example: https://cloud.google.com/appengine/docs/standard/java/datastore/transactions#uses_for_transactions

I know we haven't really discussed transactions that much but it helps prevent some race conditions that might be hard to detect. Feel free to add an issue for this and resolve in a later PR.

Reply:
OK I created a new issue for this.